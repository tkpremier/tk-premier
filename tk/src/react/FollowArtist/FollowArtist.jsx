import React, { useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import { useSelector, useDispatch } from 'react-redux';
import classNames from 'classnames';
import isNull from 'lodash/isNull';
import { saveMaker, deleteMaker } from '../PhillipsUser/actions';
import matchError from '../PhillipsUser/matchError';
import { user as userPropTypes } from '../PropTypes/proptypes';
import getPhillipsBackboneProperty from '../utils/getPhillipsBackboneProperty';
import sendAnalytics from '../../utils/sendAnalytics';

const FollowArtist = ({ makerId, makerName }) => {
  const bbEvent = useRef(null);
  useEffect(() => {
    getPhillipsBackboneProperty('Events')
      .then(eventEmitter => (bbEvent.current = eventEmitter))
      .catch(() => (bbEvent.current = null));
  }, []);
  const { error, followedMakers, user } = useSelector(({ followedMakers, user, error }) => ({
    error: matchError('FOLLOW_MAKER', maker => maker.makerId === makerId)(error) ? error : null,
    followedMakers,
    user
  }));
  const dispatch = useDispatch();
  const active = followedMakers.indexOf(makerId) > -1;
  const toolTipMessage = isNull(error) ? (active ? 'Followed' : 'Follow') : error.message;
  const trackActivity = () => {
    const page = document ? document.title : 'placeholder';
    const analyticsData = {
      eventCategory: `Follow Maker / ${page} / Maker: ${makerName}`,
      eventAction: 'Follow Maker',
      eventLabel: `Email: ${user.email}`
    };
    sendAnalytics(analyticsData);
  };
  const onClick = () => {
    if (!user.loggedIn) {
      if (!isNull(bbEvent.current)) {
        bbEvent.current.trigger('openRegister');
      }
      if (typeof localStorage !== 'undefined') {
        localStorage.setItem('followedMaker', makerId);
      }
      return;
    }
    if (active) {
      dispatch(deleteMaker(user.id, makerId));
    } else {
      trackActivity();
      dispatch(saveMaker(user.id, makerId));
    }
  };
  return (
    <button
      className={classNames('follow-artist', { active })}
      type="button"
      tabIndex={0}
      onClick={onClick}
      data-artist-id={makerId}
    >
      <span className={classNames('icon', { active })} />
      <span className={classNames('tooltip', { error: !isNull(error) })}>{toolTipMessage}</span>
    </button>
  );
};

FollowArtist.propTypes = {
  makerId: PropTypes.number.isRequired,
  makerName: PropTypes.string.isRequired
};

export { FollowArtist as UnWrappedFollowArtist };
export default FollowArtist;
