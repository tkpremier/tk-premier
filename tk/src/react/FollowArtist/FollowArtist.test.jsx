import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';
import toJson from 'enzyme-to-json';
import { UnWrappedFollowArtist } from './FollowArtist';


describe('FollowArtist Button', () => {
  const mockUser = { email: 'FakeyMcFakeUser@fake.com', id: '0' };

  Enzyme.configure({ adapter: new Adapter() });
  it('Matches the snapshot', () => {
    const component = shallow(
      <UnWrappedFollowArtist
        active={false}
        makerId={1}
        makerName="Bobson Dugnutt"
        saveMaker={() => 'mock'}
        deleteMaker={() => 'mock'}
      />
  );
    expect(toJson(component)).toMatchSnapshot();
  });

  it('when active it should have the active class', () => {
    const component = shallow(
      <UnWrappedFollowArtist
        active
        makerId={1}
        makerName="Bobson Dugnutt"
        saveMaker={() => 'mock'}
        deleteMaker={() => 'mock'}
      />
    );
    expect(component.hasClass('active')).toEqual(true);
  });

  it('when inactive onClick calls props.saveMaker', () => {
    const saveMaker = jest.fn();

    const component = shallow(
      <UnWrappedFollowArtist
        active={false}
        makerId={1}
        makerName="Bobson Dugnutt"
        saveMaker={saveMaker}
        deleteMaker={() => 'mock'}
        user={mockUser}
        userLoggedIn
      />
    );

    component.simulate('click');
    expect(saveMaker).toHaveBeenCalled();
  });

  it('when active onClick calls props.deleteMaker', () => {
    const deleteMaker = jest.fn();

    const component = shallow(
      <UnWrappedFollowArtist
        active
        makerId={1}
        makerName="Bobson Dugnutt"
        saveMaker={() => 'mock'}
        deleteMaker={deleteMaker}
        user={mockUser}
        userLoggedIn
      />
    );

    component.simulate('click');
    expect(deleteMaker).toHaveBeenCalled();
  });
});
