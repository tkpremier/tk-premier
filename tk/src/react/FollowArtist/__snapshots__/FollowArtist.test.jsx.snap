// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FollowArtist Button Matches the snapshot 1`] = `
<div
  className="follow-artist"
  data-artist-id={1}
  onClick={[Function]}
  role="button"
  tabIndex={0}
>
  <span
    className="icon"
  />
  <span
    className="tooltip"
  >
    Follow
  </span>
</div>
`;
